tools/rustfmt/srcs/lib.rs

```diff
@@ -0,0 +1,68 @@
+use std::collections::VecDeque;
+use std::env;
+use std::fs;
+use std::path::{Path, PathBuf};
+
+/// Generate an absolute path to a file without resolving symlinks
+fn absolutify_existing<T: AsRef<Path>>(path: &T) -> std::io::Result<PathBuf> {
+    let absolute_path = if path.as_ref().is_absolute() {
+        path.as_ref().to_owned()
+    } else {
+        std::env::current_dir()
+            .expect("Failed to get working directory")
+            .join(path)
+    };
+    std::fs::metadata(&absolute_path).map(|_| absolute_path)
+}
+
+/// A struct containing details used for executing rustfmt.
+#[derive(Debug)]
+pub struct RustfmtConfig {
+    /// The rustfmt binary from the currently active toolchain
+    pub rustfmt: PathBuf,
+
+    /// The rustfmt config file containing rustfmt settings.
+    /// https://rust-lang.github.io/rustfmt/
+    pub config: PathBuf,
+}
+
+/// Parse command line arguments and environment variables to
+/// produce config data for running rustfmt.
+pub fn parse_rustfmt_config() -> RustfmtConfig {
+    RustfmtConfig {
+        rustfmt: absolutify_existing(&env!("RUSTFMT")).expect("Unable to find rustfmt binary"),
+        config: absolutify_existing(&env!("RUSTFMT_CONFIG"))
+            .expect("Unable to find rustfmt config file"),
+    }
+}
+
+/// A struct of target specific information for use in running `rustfmt`.
+#[derive(Debug)]
+pub struct RustfmtManifest {
+    /// The Rust edition of the Bazel target
+    pub edition: String,
+
+    /// A list of all (non-generated) source files for formatting.
+    pub sources: Vec<String>,
+}
+
+/// Parse rustfmt flags from a manifest generated by builds using `rustfmt_aspect`.
+pub fn parse_rustfmt_manifest(manifest: &Path) -> RustfmtManifest {
+    let content = fs::read_to_string(manifest).expect(&format!(
+        "Failed to read rustfmt manifest: {}",
+        manifest.display()
+    ));
+
+    let mut lines: VecDeque<String> = content
+        .split("\n")
+        .into_iter()
+        .map(|s| s.to_owned())
+        .collect();
+
+    RustfmtManifest {
+        edition: lines
+            .pop_front()
+            .expect("There should always be at least 1 line in the manifest"),
+        sources: lines.into(),
```

Comment:
This doesn't really matter, and maybe feels super micro-optimisationy, but... If the format of the manifest file were to have the edition as the last line, rather than the first, you could use a `Vec` above, `pop().expect()`-off the edition, and avoid this `into()` :) But I definitely do not care if you ignore this comment.